source/PluginDev/Assets/GooglePlayGames/Platforms/Android/AndroidRealTimeMultiplayerClient.cs

```diff
@@ -34,111 +57,692 @@ public AndroidRealTimeMultiplayerClient(AndroidJavaObject account)
+                            mListener = listener;
+
+                            using (var task = mRtmpClient.Call<AndroidJavaObject>("join", mRoomConfig))
+                            {
+                                AndroidTaskUtils.AddOnFailureListener(
+                                    task,
+                                    e =>
+                                    {
+                                        listener.OnRoomConnected(false);
+                                        CleanSession();
+                                    });
+                            }
+                        }
+                    });
+            }
         }
 
         public void SendMessageToAll(bool reliable, byte[] data)
         {
-            // Task<Void> sendUnreliableMessageToOthers(@NonNull byte[] messageData, @NonNull String roomId)
+            if (reliable)
+            {
+                List<Participant> participants = GetConnectedParticipants();
+                foreach (Participant participant in participants)
+                {
+                    SendMessage( /* reliable= */ true, participant.ParticipantId, data);
+                }
+
+                return;
+            }
+
+            var roomStatus = GetRoomStatus();
+            if (roomStatus != RoomStatus.Active && roomStatus != RoomStatus.Connecting)
+            {
+                OurUtils.Logger.d("Sending message is not allowed in this state.");
+                return;
+            }
+
+            string roomId = mRoom.Call<string>("getRoomId");
+            using (mRtmpClient.Call<AndroidJavaObject>("sendUnreliableMessageToOthers", data, roomId)) ;
         }
 
         public void SendMessageToAll(bool reliable, byte[] data, int offset, int length)
         {
-            // Task<Void> sendUnreliableMessageToOthers(@NonNull byte[] messageData, @NonNull String roomId)
+            SendMessageToAll(reliable, Misc.GetSubsetBytes(data, offset, length));
         }
 
         public void SendMessage(bool reliable, string participantId, byte[] data)
         {
-            // Task<Integer> sendReliableMessage(@NonNull byte[] messageData, @NonNull String roomId, @NonNull String recipientParticipantId, @Nullable ReliableMessageSentCallback callback)
-            // Task<Void> sendUnreliableMessage(@NonNull byte[] messageData, @NonNull String roomId, @NonNull String recipientParticipantId)
+            var roomStatus = GetRoomStatus();
+            if (roomStatus != RoomStatus.Active && roomStatus != RoomStatus.Connecting)
+            {
+                OurUtils.Logger.d("Sending message is not allowed in this state.");
+                return;
+            }
+
+            if (GetParticipant(participantId) == null)
+            {
+                OurUtils.Logger.e("Attempted to send message to unknown participant " + participantId);
+                return;
+            }
+
+            string roomId = mRoom.Call<string>("getRoomId");
```

Comment:
Seems as we should cache roomId and not get it for each call

Reply:
This line is used in LeaveRoom and in SendMessage. For the first one, caching doesn't decrease the cost and for the second one, it doesn't cost too much as we can have only 3 players to SendMessage. 

On the other hand, there are more than places to change mRoom value. Caching it won't have much effect on cost but it will increase code complexity. That's why, I'd rather keep it this way.